src/perform_step/exponential_rk_perform_step.jl

```diff
@@ -1158,6 +1158,92 @@ function perform_step!(integrator, cache::EPIRK5P2Cache, repeat_step=false)
   # integrator.k is automatically set due to aliasing
 end
 
+######################################################
+# Adaptive exponential Rosenbrock integrators
+function initialize!(integrator, cache::Exprb32ConstantCache)
+  # Pre-start fsal
+  integrator.fsalfirst = integrator.f(integrator.uprev, integrator.p, integrator.t)
+  integrator.fsallast = zero(integrator.fsalfirst)
+
+  # Initialize interpolation derivatives
+  integrator.kshortsize = 2
+  integrator.k = typeof(integrator.k)(undef, integrator.kshortsize)
+  integrator.k[1] = integrator.fsalfirst
+  integrator.k[2] = integrator.fsallast
+end
+function initialize!(integrator, cache::Exprb32Cache)
+  # Pre-start fsal
+  integrator.fsalfirst = zero(cache.rtmp)
+  integrator.f(integrator.fsalfirst, integrator.uprev, integrator.p, integrator.t)
+  integrator.fsallast = zero(integrator.fsalfirst)
+
+  # Initialize interpolation derivatives
+  integrator.kshortsize = 2
+  resize!(integrator.k, integrator.kshortsize)
+  integrator.k[1] = integrator.fsalfirst
+  integrator.k[2] = integrator.fsallast
+end
+
+function perform_step!(integrator, cache::Exprb32ConstantCache, repeat_step=false)
+  @unpack t,dt,uprev,f,p = integrator
+  is_compos = isa(integrator.alg, CompositeAlgorithm)
+  A = isa(f, SplitFunction) ? f.f1 : calc_J!(integrator,cache,is_compos) # get linear operator
+  alg = unwrap_alg(integrator, true)
+
+  F1 = integrator.fsalfirst
+  w1 = phiv(dt, A, F1, 3; m=min(alg.m, size(A,1)), opnorm=integrator.opts.internalopnorm, iop=alg.iop)
+  U2 = uprev + dt * w1[:, 2]
+  F2 = _compute_nl(f, U2, p, t + dt, A) + A * uprev
+  w2 = phiv(dt, A, F2, 3; m=min(alg.m, size(A,1)), opnorm=integrator.opts.internalopnorm, iop=alg.iop)
+  u = uprev + dt * (w1[:,2] - 2w1[:,4] + 2w2[:,4])
+  if integrator.opts.adaptive
+    utilde = dt * w1[:,2] # embedded method
```

Comment:
In the language of the paper, this is `uhat`, right? So then in our setup, the error estimate is the first argument of `calculate_residuals` which should be `u - utilde`. However, it's more numerically stable to never construct `uhat` itself. Instead, define `utilde = u - uhat = [-2phi_3 phi_3] = 2dt*(-w1[:,4] + w2[:,4])` as the error estimator.